content/article/creating-midi-music-with-perl.md

```diff
@@ -0,0 +1,443 @@
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-1.mp3" type="audio/mpeg" >}}
+
+Rendering Audio
+---------------
+
+In order to actually hear some sound, you can either play the MIDI directly, with a command-line player like `timidity` and an `sf2` "soundfont" file, as in [this list](https://archive.org/download/free-soundfonts-sf2-2019-04). Also the MIDI file can be used to create an audio formatted file (e.g. WAV, MP3) that can be played. Here is the command for this, that I use on my Mac:
+
+```
+timidity -c ~/timidity.cfg some.mid -Ow -o - | ffmpeg -i - -acodec libmp3lame -ab 64k some.mp3
+```
+
+But wait! You can also generate and play MIDI in real-time with the [MIDI::RtMidi::FFI::Device]({{< mcpan "MIDI::RtMidi::FFI::Device" >}}) and [MIDI::RtMidi::ScorePlayer]({{< mcpan "MIDI::RtMidi::ScorePlayer" >}}) modules. :D
+
+Back to Creating Music!
+-----------------------
+
+So far, we have encountered the "Setup, Play, and Write" algorithm. Next we shall replace the "Play" bit with "Sync" and play the bass and treble parts simultaneously:
+
+```perl
+use MIDI::Util qw(setup_score);
+
+my $score = setup_score();
+
+$score->synch(
+  sub { bass($score) },
+  sub { treble($score) },
+);
+
+$score->write_score("$0.mid");
+
+sub bass {
+  my ($score) = @_;
+  for my $note (qw(C3 F3 G3 C4)) {
+    $score->n('qn', $note);
+  }
+}
+
+sub treble {
+  my ($score) = @_;
+  for my $note (qw(C4 D4 E4 F4)) {
+    $score->n('en', $note);
+    $score->r('en');
+  }
+}
+```
+
+This algorithm is not especially clever, but illustrates the basics.
+
+If we want to repeat the phrase, just add a `for` loop to the `synch`:
+
+```perl
+$score->synch(
+  sub { bass($score) },
+  sub { treble($score) },
+) for 1 .. 4;
+```
+
+{{< audio src="/images/creating-midi-music-with-perl/audio-2.mp3" type="audio/mpeg" >}}
+
+Setting Channels, Patches, Volume, and Tempo
+--------------------------------------------
+
+Use the `MIDI::Util::set_chan_patch()` function to set the channel and the patch.
+
+To set the tempo in "beats per minute", give a `bpm` to the `setup_score()` function. To set an individual note volume, add `"v$num"` as an argument to the `$score->n()` method, where `$num` is an integer from 0 to 127. Here is code that sets these parameters:
+
+```perl
+use MIDI::Util qw(setup_score set_chan_patch);
+
+my $bpm = shift || 120;
+
+my $score = setup_score(bpm => $bpm);
+
+$score->synch(\&bass, \&treble);
+
+$score->write_score("$0.mid");
+
+sub bass {
+  set_chan_patch($score, 0, 35);
+  for my $note (qw(C3 F3 G3 C4)) {
+    $score->n('qn', $note, 'v127');
+  }
+}
+
+sub treble {
+  set_chan_patch($score, 1, 0);
+  for my $note (qw(C4 D4 E4 F4)) {
+    $score->n('en', $note, 'v110');
+    $score->r('en');
+  }
+}
+```
+
+Notice that the `\&` syntax is used for the `synch`'d subs. This makes the bass and treble subroutines use the `$score` object as a global variable. Sometimes this is done... Some may hate this. Fortunately, it's optional.
+
```

Comment:
What happens if you try to synch two subs which do not produce notes which add up to the same amount?

Reply:
I am not sure what you mean, but I will remove this extra, confusing, unnecessary bit about code-refs.

Reply:
Oh. I understand. The `synch` will be jumbled-up and the parts will not align. I will think about how to say this...

Reply:
Ok fixed. Hopefully my change explains things a bit better.